### PR TITLE
Implement DeepONet Architecture

### PR DESCRIPTION
**Pull request by AshwinSaklecha**
## Description

This PR adds DeepONet, a neural network architecture for learning operators that map between input and output functions. The implementation includes the base model, a DeepChem wrapper, and unit tests.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/tests/test_deeponet.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    has_torch = True
+except ModuleNotFoundError:
+    has_torch = False
+
+
+@pytest.mark.torch
+@pytest.mark.skipif(not has_torch, reason="PyTorch is not installed")
+def test_deeponet_construction():
+    """Test that DeepONet base model can be constructed with correct attributes."""
+    from deepchem.models.torch_models.deeponet import DeepONet
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/tests/test_deeponet.py]: Can you please call the `DeepONet` and `DeepONetModel` from `torch_models` since you have alredy added them to `__init__.py` file?

This should be something like `from deepchem.models.torch_models import DeepONet`

File: deepchem/models/torch_models/tests/test_deeponet.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    """Test that DeepONetModel can fit to data and predict correct shape."""
+    from deepchem.models.torch_models import DeepONetModel
+
+    branch_input_dim = 10
+    trunk_input_dim = 3
+    n_samples = 20
+
+    # Create concatenated input: X = [branch_data | trunk_data]
+    branch_data = np.random.randn(n_samples,
+                                  branch_input_dim).astype(np.float32)
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/tests/test_deeponet.py:76]: Is there an especial reason to making this `.astype(np.float32)`?

File: deepchem/models/torch_models/tests/test_deeponet.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+                          batch_size=5)
+    model.fit(dataset, nb_epoch=1)
+
+    pred = model.predict_on_batch(X)
+    assert pred.shape == y.shape
+
+
+@pytest.mark.torch
+@pytest.mark.skipif(not has_torch, reason="PyTorch is not installed")
+def test_deeponetmodel_restore():
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/tests/test_deeponet.py:94]: Can you please also test for model saving?

File: deepchem/models/torch_models/tests/test_deeponet.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+@pytest.mark.skipif(not has_torch, reason="PyTorch is not installed")
+def test_deeponetmodel_overfit():
+    """Test that DeepONetModel can overfit on a small dataset."""
+    from deepchem.models.torch_models import DeepONetModel
+
+    branch_input_dim = 5
+    trunk_input_dim = 2
+    n_samples = 10
+
+    np.random.seed(42)
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/tests/test_deeponet.py]: Can you please avoid using random seed? This is a minor detail, but some tests can be flaky and we should be able to detect them.

File: deepchem/models/torch_models/tests/test_deeponet.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+                          trunk_input_dim=trunk_input_dim,
+                          branch_hidden=(64, 64),
+                          trunk_hidden=(64, 64),
+                          output_dim=64,
+                          batch_size=10,
+                          learning_rate=1e-3)
+    model.fit(dataset, nb_epoch=500)
+
+    pred = model.predict_on_batch(X)
+    mse = np.mean((pred - y)**2)
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/tests/test_deeponet.py]: Can you please use metrics of DeepChem's API? 


`deepchem.metrics.Metric(dc.metrics.mean_squared_error, mode='regression')`

**Comment by JoseAntonioSiguenza** [comment]
@AshwinSaklecha Nice work! Some details to be addressed and check CI.

File: deepchem/models/torch_models/tests/test_deeponet.py (hunk 2 above)
Comment (AshwinSaklecha) [reply deepchem/models/torch_models/tests/test_deeponet.py:76]: I used np.random.randn() which returns float64 by default, and pytorch expects float32, so to prevent mismatching in datatypes, typecasting was needed. This same pattern is used in other test files like 
test_hnn.py too.

File: deepchem/models/torch_models/tests/test_deeponet.py (hunk 3 above)
Comment (AshwinSaklecha) [reply deepchem/models/torch_models/tests/test_deeponet.py:94]: Actually I have already covered it in test_deeponetmodel_restore function only. In this test, the model is saved after training, then restored in a new instance, and predictions are verified to match.

File: deepchem/models/torch_models/tests/test_deeponet.py (hunk 5 above)
Comment (AshwinSaklecha) [reply deepchem/models/torch_models/tests/test_deeponet.py]: Done, updated.

File: deepchem/models/torch_models/tests/test_deeponet.py (hunk 4 above)
Comment (AshwinSaklecha) [reply deepchem/models/torch_models/tests/test_deeponet.py]: Done.

File: deepchem/models/torch_models/tests/test_deeponet.py (hunk 1 above)
Comment (AshwinSaklecha) [reply deepchem/models/torch_models/tests/test_deeponet.py]: Done.

File: deepchem/models/torch_models/tests/test_deeponet.py (hunk 3 above)
Comment (JoseAntonioSiguenza) [reply deepchem/models/torch_models/tests/test_deeponet.py:94]: Sure. But we need to make sure that `.save` method which `DeepONetModel` inherits from `TorchModel` works properly.

Comment (AshwinSaklecha) [reply deepchem/models/torch_models/tests/test_deeponet.py:94]: Done, I have added model.save_checkpoint method to explicitly test the save functionality before restore.

**Comment by JoseAntonioSiguenza** [comment]
Nice work, LGTM, lets see how CI behaves.

File: docs/source/api_reference/models.rst [hunk 6]
```diff
@@ -671,6 +671,16 @@ LNNModel
 .. autoclass:: deepchem.models.torch_models.LNNModel
   :members:
 
+DeepONet
```
Comment (rbharath) [inline docs/source/api_reference/models.rst:674]: Can you add this to the model cheatsheet? @JoseAntonioSiguenza can point you there

Comment (AshwinSaklecha) [reply docs/source/api_reference/models.rst:674]: Done, I have added DeepONetModel to general_purpose_models.csv file along with the DeepONet paper reference.

**Comment by rbharath** [comment]
@JoseAntonioSiguenza I am seeing a lot of failures on simdna and pkg_resources? I think these are not related to the current PR; are you seeing these elsewhere?

**Comment by JoseAntonioSiguenza** [comment]
> @JoseAntonioSiguenza I am seeing a lot of failures on simdna and pkg_resources? I think these are not related to the current PR; are you seeing these elsewhere?

Yes, these breakages are not related to this PR. I've seen these failures in other PRs and I'm looking into them (I believe is due to version mismatch)
